### ISSUE TITLE
[Bug]: New page event handler doesn't catch first request in new tab

### ISSUE THREAD

**Issue by marshmn** [microsoft/playwright#30229]
### Version

1.42.1

### Steps to reproduce

I've created an override for the default `context` fixture, which adds an event handler for the `page` event. The handler adds an additional header to any requests made via the new page.

This seems to work for most scenarios, such as:
- Tests using the normal `page` fixture
- New pages created explicitly from a call to `context.newPage()`

However, in a scenario where a new page is created through clicking on a link with a `target="_blank"` attribute, the event handler doesn't appear to fire fast enough to catch the first request to the new page. It does appear to have fired by the time subsequent requests to that same new page are made.

Here are some example scenarios:
```typescript
import { test as base } from "@playwright/test";

export const test = base.extend({
    context: async ({ context }, use) => {
        context.on("page", async (page) => {
            await page.route(
                new RegExp(".*"),
                (route, request) => {
                    const headers = {
                        ...request.headers(),
                        "X-Foo": "bar",
                    };
                    void route.continue({ headers });
                },
            );
        });
        await use(context);
    },
});

test("scenario 1", async ({ page }) => {
    // Use page fixture
    // SUCCESS: header appears in request as expected
    await page.goto("/");
});

test("scenario 2", async ({ context }) => {
    // Create page object from context
    // SUCCESS: header appears in request as expected
    const page = await context.newPage();
    await page.goto("/");
});

test("scenario 3", async ({ page, context }) => {
    await page.goto("/");

    // Click link with `target="_blank"` attribute to open new tab
    // FAIL: header does **not** appear in request on new page object (newPage)
    const pagePromise = context.waitForEvent("page");
    await page.getByRole("link", { name: "Link" }).click();
    const newPage = await pagePromise;
    await newPage.waitForLoadState();

    // SUCCESS: header appears in subsequent requests using same page object
    await newPage.goto("/");
});
```

I've also created an example repo with this code: https://github.com/marshmn/pw-new-page-event-test

### Expected behavior

I would expect the `X-Foo: bar` header to appear in all requests.

### Actual behavior

In the 3rd scenario above, there are 3 requests - the header appears in the first and third request, but not in the second request (which is the one which actually initiates the new page via click on link with `target="_blank"` attribute.

### Additional context

I've collected a trace file for scenario 3: 
[trace.zip](https://github.com/microsoft/playwright/files/14858166/trace.zip)

In that trace, the network tab shows the first and third requests with blue background - the second with white (I'm not sure what this signifies, but I'm guessing it has some relevance...).

![Screenshot from 2024-04-03 21-11-54](https://github.com/microsoft/playwright/assets/4307033/8447c3d0-dc05-430c-b475-ba3cd227a100)

### Environment

```shell
System:
  OS: Linux 5.11 Debian GNU/Linux 12 (bookworm) 12 (bookworm)
  CPU: (12) x64 Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
  Memory: 11.63 GB / 31.00 GB
  Container: Yes
Binaries:
  Node: 20.12.0 - /usr/local/bin/node
  Yarn: 1.22.19 - /usr/local/bin/yarn
  npm: 10.5.0 - /usr/local/bin/npm
Languages:
  Bash: 5.2.15 - /usr/bin/bash
npmPackages:
  @playwright/test: ^1.42.1 => 1.42.1
```

**Comment by dgozman** [comment]
@marshmn This works as expected. Browser issues a request first, and only after it gets a response, it would create a page. There is no way to route the request on the `page` instance for the popup. However, you should use `context.route()` instead - this one will catch all the requests including the popup main request. Let me know whether this helps.

**Comment by marshmn** [comment]
@dgozman: Thanks for the explanation.

I've switched to use `context.route()` and all now appears to work as I'd like. It's also a slightly neater solution, so I'm happy with this - thanks for pointing it out. 

For anyone else who stumbles across this thread later, here's how I changed my fixture in the above code:
```typescript
export const test = base.extend({
    context: async ({context}, use) => {
        await context.route(
            new RegExp(".*"),
            (route, request) => {
                const headers = {
                    ...request.headers(),
                    "X-Foo": "bar",
                };
                void route.continue({headers});
            },
        );
        await use(context);
    },
});
```

Perhaps the documentation, for either [`BrowserContext.on('page')`](https://playwright.dev/docs/api/class-browsercontext#browser-context-event-page) and/or [Handling new pages](https://playwright.dev/docs/pages#handling-new-pages), should be improved to indicate what to expect regarding new pages/popups with the `page` event? Even though what I was doing turned out to be expected behaviour, I don't think that I could have known that from reading the docs?

Sorry for taking up so much of your time with my various queries - but thanks again for your help in understanding what I needed to do here.
